core/test/database/MemoryDatabaseProxy.cpp

```diff
@@ -357,18 +362,105 @@ class Connection : public api::DatabaseConnection {
         return std::make_shared<Blob>();
     }
 
+    void chech_sqlite_err(sqlite3 *db, int result, const std::string &errorMessage) {
+        if (SQLITE_OK != result)
+        {
+            const char *errMsg = sqlite3_errmsg(db);
+            std::ostringstream ss;
+            ss << errorMessage << errMsg;
+            throw make_exception(api::ErrorCode::RUNTIME_ERROR, ss.str());
+        }
+    };
+
+    void changePassword(const std::string & oldPassword, const std::string & newPassword) {
+        setPassword(oldPassword);
+        if (!newPassword.empty()) {
+#ifdef SQLCIPHER
+            auto res = sqlite3_rekey_v2(_db, _dbName.c_str(), newPassword.c_str(), strlen(newPassword.c_str()));
+            chech_sqlite_err(_db, res, "Failed to change database's password. ");
+#endif
+        }
+    };
+
+    void setPassword(const std::string &password) {
+        if (!password.empty()) {
+#ifdef SQLCIPHER
+            auto res = sqlite3_key_v2(_db, _dbName.c_str(), password.c_str(), strlen(password.c_str()));
+            chech_sqlite_err(_db, res, "Failed to encrypt database. ");
+#endif
+        }
+    };
+
 private:
     sqlite3* _db;
+    std::string _dbName;
+};
+
+// Parse parameters got from opening a session
+// Same as in soci_sqlite
+static std::tuple<std::string, std::string, std::string> parseParameters(const std::string &parameters) {
+
+    std::string passKey, newPassKey;
+    std::string dbname(parameters);
+    std::stringstream ssconn(parameters);
+    while (!ssconn.eof() && ssconn.str().find('=') != std::string::npos)
```

Comment:
Well it's normally set up to max value of `size_t`, but I don't know about `c_end` is it on `stringstream` ?

Reply:
No, but when you do a `str().find('=')`, you’re searching in a `std::string`, so you could store that string and use its `cend()` method (it’s `cend` not `c_end`, I was mistaken). But since you use `std::string::find` and not `std::find`, I guess you’re fine (even though I would have expected an iterator and not a `size_t`, but that’s fine I guess, especially for your test).